### PR TITLE
Fix build failure on riscv64 architecture

### PR DESCRIPTION
**Pull request by imguoguo**
When attempting to compile on a riscv64 machine, the compilation process fails with an error message indicating that the "march=native" option is not supported on the riscv64 architecture. This option is used to automatically detect and optimize the compilation for the native architecture of the machine. However, since it is not compatible with riscv64, it leads to a compilation failure.

To resolve this issue(#3874), this pull request proposes a fix by manually specifying the appropriate architecture flag for riscv64.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3875](https://github.com/mfem/mfem/pull/3875) | @imguoguo | @tzanio | @sebastiangrimberg + @tzanio | 9/19/23 | 11/15/23 | 11/19/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by imguoguo** [comment]
I have successfully compiled it using 'rpmbuild -ba mfem.spec' in the [mfem.spec](https://github.com/fedora-riscv/mfem/blob/f38-rv64/mfem.spec) file under riscv64 qemu virtual machine, and all tests have passed.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by tzanio — APPROVED** [review]

File: miniapps/performance/makefile [hunk 1]
```diff
@@ -98,9 +98,13 @@ MFEM_PERF_CXXFLAGS_gcc_ppc = -mcpu=native -mtune=native\
 MFEM_PERF_CXXFLAGS_xlc = -mcpu=native
 
 # - Clang extra options:
-ifneq ($(MFEM_MACHINE),arm64)
+ifeq ($(MFEM_MACHINE),riscv64)
+  MFEM_PERF_CXXFLAGS_clang += -march=rv64gc
+  else
+    ifneq ($(MFEM_MACHINE),arm64)
```
Comment (sebastiangrimberg) [inline miniapps/performance/makefile:104]:
```suggestion
else ifneq ($(MFEM_MACHINE),arm64)
```

**Review by sebastiangrimberg — CHANGES_REQUESTED** [review]

File: miniapps/performance/makefile [hunk 2]
```diff
@@ -98,9 +98,13 @@ MFEM_PERF_CXXFLAGS_gcc_ppc = -mcpu=native -mtune=native\
...
   # -march=native is unavailable on clang/ARM64 as of 05/2021: support could be added later.
```
Comment (sebastiangrimberg) [inline miniapps/performance/makefile:105]: Does this still apply? Seems like it works OK for the CMake build.

**Comment by tzanio** [comment]
ping: @imguoguo

File: miniapps/performance/makefile (hunk 2 above)
Comment (imguoguo) [reply miniapps/performance/makefile:104]: After testing building on riscv64, these lines are not necessary and can be removed, only the modification of CMakeList.txt is necessary.

**Comment by sebastiangrimberg** [comment]
@imguoguo I am a bit confused by your latest commit. If no changes are required to the GNU Make build for this to work, then I would expect none are required for CMake and this PR can be closed.

On the other hand, if the CMake changes are required I think you need to make the corresponding Makefile changes to be consistent across different build systems.

Which is the case or am I misunderstanding?

**Comment by imguoguo** [comment]
> @imguoguo I am a bit confused by your latest commit. If no changes are required to the GNU Make build for this to work, then I would expect none are required for CMake and this PR can be closed.
> 
> On the other hand, if the CMake changes are required I think you need to make the corresponding Makefile changes to be consistent across different build systems.
> 
> Which is the case or am I misunderstanding?

When you said "Does this still apply," I mistakenly thought you were asking if my modification was effective. Therefore, I attempted to remove the changes made to the makefile and found that indeed, it is possible to compile without modifying the makefile. However, the modifications to the CMakeLists.txt file are still necessary. As a result, I conducted several compilation tests.

After multiple building tests on a riscv64 machine, I discovered that the previous commit's revert was due to the fact that this issue does not occur under GNU Make. It only manifests when using CMake. By making modifications to the CMakeLists.txt file, I was able to successfully compile and test the code. However, if I do not make these modifications, the error message(g++: error: '-march=native': ISA string must begin with rv32 or rv64
) will appear during building. In order to maintain consistency and avoid future complications, I suggest modifying both files in a subsequent commit.

I apologize for any inconvenience caused by my delayed response during this busy period.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**